File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
sys.setrecursionlimit(10**9)
input = sys.stdin.readline

def dfs(ox,oy) : 
  global cnt, vNum
  cx = [1, -1, 0, 0]
  cy = [0, 0, 1, -1]

  if ox < 0 or ox >= n or oy < 0 or oy >= n:
        return False

  if board[ox][oy] == 1:
    vNum +=1 
    board[ox][oy] = 2
    for i in range(4):
        X = ox + cx[i]
        Y = oy + cy[i]
        dfs(X,Y)


n = int(input())
board = [[0] * (n) for _ in range(n)] 
vNum = 0
cnt = 0
vNumList = []

for i in range(n) : 
    line = list(map(int, input().strip()))
    for j in range(n) : 
       board[i][j] = line[j] 

for i in range(n) : 
    for j in range(n) : 
        if board[i][j] == 1 :
            dfs(i,j)
            cnt += 1
            vNumList.append(vNum)
            vNum = 0


print(cnt)   
vNumList.sort()
for i in vNumList :
    print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const N = +input.shift();
let board = input.map((item) => item.split("").map(Number));

let dx = [0, 0, 1, -1];
let dy = [-1, 1, 0, 0];

let cnt = 0;
let answer = [];

// let visited = [...new Array(N)].map((v) => new Array(N).fill(false));

function dfs(x, y) {
  board[x][y] = 0;
  cnt++;
  for (let i = 0; i < 4; i++) {
    let X = x + dx[i];
    let Y = y + dy[i];
    if (X >= 0 && X < N && Y >= 0 && Y < N && board[X][Y] !== 0) {
      dfs(X, Y);
    }
  }
}

for (let i = 0; i < N; i++) {
  for (let j = 0; j < N; j++) {
    if (board[i][j] == 1) {
      dfs(i, j);
      answer.push(cnt);
      cnt = 0;
    }
  }
}

console.log(answer.length);
console.log(answer.sort((a, b) => a - b).join("\n"));
